### DISCUSSION TITLE
questions related to --pack-dyn-relocs=android

### DISCUSSION THREAD

**Discussion by pendyalasyam** [android/ndk#1800] [Q&A] [answered]
Build System maintainers [guide](https://android.googlesource.com/platform/ndk/+/master/docs/BuildSystemMaintainers.md) suggests to use **--pack-dyn-relocs=android** on API Level 23 and more. 

I am trying to understand what are relocations and what is this new packing and how it reduces the binary size. Meanwhile, could somebody please let me know

1. What happens if we use libraries built with **--pack-dyn-relocs=android** on Android Devices with lower API levels. 
2. Can libraries built with **--pack-dyn-relocs=android** work smoothly with libraries that are built without any --pack-dyn-relocs flag?
3. Does this flag impact the performance of application by any chance? If so, please let me know why?

Thanks!!

**Comment by enh-google** [comment, marked as answer]
> Build System maintainers [guide](https://android.googlesource.com/platform/ndk/+/master/docs/BuildSystemMaintainers.md) suggests to use **--pack-dyn-relocs=android** on API Level 23 and more.
> 
> I am trying to understand what are relocations and what is this new packing and how it reduces the binary size. Meanwhile, could somebody please let me know
> 
> 1. What happens if we use libraries built with **--pack-dyn-relocs=android** on Android Devices with lower API levels.

they won't work.

more specifically: https://android.googlesource.com/platform/bionic/+/master/android-changes-for-ndk-developers.md#relative-relocations-relr

> 2. Can libraries built with **--pack-dyn-relocs=android** work smoothly with libraries that are built without any --pack-dyn-relocs flag?

yes.

> 3. Does this flag impact the performance of application by any chance? If so, please let me know why?

no. (it might _reduce_ the time taken to dlopen() the library, but once it's loaded, it will run exactly the same as before.)

**Reply by pendyalasyam** [reply]
Thanks @enh-google !!
